### PR TITLE
xapi-idl: Delete String.{explode,implode} functions

### PR DESCRIPTION
**Pull request by psafont**
These are highly inefficient.

Also changes some functions to be able to have less types and make normal usage
clearer. This comes at the cost of having to destructure the main type when
pattern-matching it.

Moves the device_number tests to its own executable to easily iterate on the
tests.

Drafting as it needs to pass tests regarding urlencoding

### CODE REVIEW

**Comment by psafont** [comment]
I've run a suite to verify that cookie parsing is not disturbed. Most of the code is unit tested (split_f and the device code in particular). I think this is ready to go

File: ocaml/idl/markdown_backend.ml [hunk 1]
```diff
@@ -43,8 +43,6 @@ let compare_case_ins x y =
   compare (String.lowercase_ascii x) (String.lowercase_ascii y)
 
 let escape s =
-  let open Xapi_stdext_std.Xstringext in
```
Comment (edwintorok) [inline ocaml/idl/markdown_backend.ml:46]: There is a project called `cmarkit` that can be use to generate Markdown from OCaml, perhaps in the future we might want to use it instead of generating markdown by hand here.
It can both parse and render Markdown https://ocaml.org/p/cmarkit/latest/doc/index.html

File: ocaml/libs/xapi-stdext/lib/xapi-stdext-std/xstringext.ml [hunk 2]
```diff
@@ -81,24 +67,28 @@ module String = struct
         in
         concat "" (fold_right aux string [])
 
-  (** Take a predicate and a string, return a list of strings separated by
-      runs of characters where the predicate was true (excluding those characters from the result) *)
   let split_f p str =
```
Comment (edwintorok) [inline ocaml/libs/xapi-stdext/lib/xapi-stdext-std/xstringext.ml:70]: Can we use https://ocaml.org/p/astring/latest/doc/Astring/String/index.html#val-fields here?

File: ocaml/xapi-idl/xen/device_number.ml [hunk 3]
```diff
@@ -76,13 +78,12 @@ let of_xenstore_int x =
         (Floppy, (x lsr 4) land ((1 lsl 4) - 1), x land ((1 lsl 4) - 1))
     | n ->
         let idx =
-          snd
-            (List.fold_left
```
Comment (edwintorok) [inline ocaml/xapi-idl/xen/device_number.ml:98]: We could build a Map in a global var that gives you the answer directly. The table is not going to change.

File: ocaml/xapi/xapi_xenops.ml [hunk 4]
```diff
@@ -555,31 +555,37 @@ module MD = struct
...
-            Lowest
-        | s -> (
-          try Other (int_of_string s)
-          with _ ->
-            warn "Unknown VBD QoS scheduler class (try 'high' 'low' 'normal')" ;
-            Normal
-        )
-      else
-        Normal
+      match List.assoc_opt "class" params with
```
Comment (edwintorok) [inline ocaml/xapi/xapi_xenops.ml:570]: These are not related to String.implode, might've been useful to have them in a separate commit.

Comment (edwintorok) [reply ocaml/xapi/xapi_xenops.ml:570]: Could you update the commit message with a description of what else to expect in this commit? (e.g. the assoc_opt optimization, and I also see some device number handling refactoring)

File: ocaml/xapi-idl/xen/device_number.ml (hunk 3 above)
Comment (psafont) [reply ocaml/xapi-idl/xen/device_number.ml:98]: I've replaced the list with an array and an intmap for faster accesses

File: ocaml/libs/xapi-stdext/lib/xapi-stdext-std/xstringext.ml (hunk 2 above)
Comment (psafont) [reply ocaml/libs/xapi-stdext/lib/xapi-stdext-std/xstringext.ml:70]: I'd rather avoid importing Astring if not needed, the function is well tested with unit tests

File: ocaml/xapi/xapi_xenops.ml (hunk 4 above)
Comment (psafont) [reply ocaml/xapi/xapi_xenops.ml:570]: I've described both changes in the commit. Let me know if it's clarifying enough

File: ocaml/xapi-idl/xen/device_number.ml [hunk 5]
```diff
@@ -119,112 +106,119 @@ let rec string_of_int26 x =
...
+    let* disk, partition = parse_b26_int rest in
+    Some (Xen, disk, partition)
+  else if String.is_prefix ~affix:"sd" x then
+    let rest = String.sub_with_range ~first:2 x in
+    let* disk, partition = parse_b26_int rest in
+    Some (Scsi, disk, partition)
+  else if String.is_prefix ~affix:"fd" x then
+    let rest = String.sub_with_range ~first:2 x in
+    let* disk, partition = parse_b26_int rest in
+    Some (Floppy, disk, partition)
```
Comment (psafont) [inline ocaml/xapi-idl/xen/device_number.ml:192]: This doesn't restrict the value to fda and fdb like `make` does, the code needs to call it

File: ocaml/xapi/xapi_vbd_helpers.ml [hunk 6]
```diff
@@ -376,38 +376,17 @@ let clear_current_operations ~__context ~self =
...
-  match String.explode dev with
-  | 's' :: 'd' :: 'a' .. 'p' :: rest ->
-      check_rest rest
-  | 'x' :: 'v' :: 'd' :: 'a' .. 'p' :: rest ->
-      check_rest rest
-  | 'h' :: 'd' :: 'a' .. 'p' :: rest ->
-      check_rest rest
-  | 'f' :: 'd' :: 'a' .. 'b' :: rest ->
-      check_rest rest
-      (* QEMU only supports up to 2 floppy drives, hence fda or fdb *)
```
Comment (psafont) [inline ocaml/xapi/xapi_vbd_helpers.ml:400]: This contradicts the limit in device_number, which allows for the staggering amount of 3 floppy drives

Comment (psafont) [reply ocaml/xapi/xapi_vbd_helpers.ml:400]: qemu's manual agrees with the comment https://www.qemu.org/docs/master/system/qemu-manpage.html

File: ocaml/xapi-idl/xen/device_number.ml [hunk 7]
```diff
@@ -119,112 +106,119 @@ let rec string_of_int26 x =
...
+    if String.Sub.is_empty x then
+      Some (pre, 0)
+    else
+      let* post, x = parse_int x in
+      if not (String.Sub.is_empty x) then
+        None
+      else
+        Some (pre, post)
   in
   (* Parse a string "123p456" into x, y where x = 123 and y = 456 *)
```
Comment (lindig) [inline ocaml/xapi-idl/xen/device_number.ml:161]: Why not use `sscanf` with `%dp%d`? A lot of this parsing code looks too complicated.

**Review by lindig — APPROVED** [review]
This module has a lot of code that looks too complicated and this is a step in the right direction - especially for functions that are used frequently. I suspect more simplification is possible when it comes to parsing and decomposing strings.

File: ocaml/xapi/xapi_dr_task.ml [hunk 8]
```diff
@@ -31,7 +31,7 @@ type sr_probe_sr = {uuid: string; name_label: string; name_description: string}
 (* Attempt to parse a key/value pair from XML. *)
 let parse_kv = function
   | Xml.Element (key, _, [Xml.PCData v]) ->
-      (key, String.strip String.isspace v) (* remove whitespace at both ends *)
+      (key, String.trim v)
```
Comment (Vincent-lau) [inline ocaml/xapi/xapi_dr_task.ml:34]: Is it a problem that we are trimming these `\x0C' (form feed), '\n', '\r', and '\t' as well as the space?

**Review by Vincent-lau — APPROVED** [review]
Original code looks quite complex, but as long as there are good test coverage this should be good

File: ocaml/xapi-idl/lib_test/device_number_test.ml [hunk 9]
```diff
@@ -64,48 +66,58 @@ let test_examples =
...
+  ("Reject invalid devices", tests)
 
 let test_2_way_convert =
   (* We now always convert Ide specs into xvd* linux devices, so they become Xen
      specs when converted back. *)
-  let equal_linux old_t new_t =
-    match (spec old_t, spec new_t) with
+  let equal_linux (old_t : t) (new_t : t) =
+    match
+      ((old_t, new_t) :> (bus_type * int * int) * (bus_type * int * int))
```
[the remark this answers is not in the document]
Comment (psafont) [reply ocaml/xapi-idl/lib_test/device_number_test.ml:119]: They are needed because the type is private

File: ocaml/libs/xapi-stdext/lib/xapi-stdext-std/xstringext.ml [hunk 10]
```diff
@@ -81,24 +67,28 @@ module String = struct
...
-    let not_p x = not (p x) in
-    let rec split_one p acc = function
-      | [] ->
-          (List.rev acc, [])
-      | c :: cs ->
-          if p c then split_one p (c :: acc) cs else (List.rev acc, c :: cs)
+    let split_one seq =
+      let not_p c = not (p c) in
+      let a = Seq.take_while not_p seq in
```
[the remark this answers is not in the document]
Comment (psafont) [reply ocaml/libs/xapi-stdext/lib/xapi-stdext-std/xstringext.ml:73]: The purpose of the whole function is to return a list of string of characters that do _not_ satisfy p, separated by strings of characters that satisfy p.

for example:
```ocaml
let p c = c = ' ' in
split_f p "this      is   a test"
- : string list = ["this"; "is"; "a"; "test"]
```

File: ocaml/xapi/xapi_dr_task.ml (hunk 8 above)
Comment (psafont) [reply ocaml/xapi/xapi_dr_task.ml:34]: isspace is a misnomer, it should be called is_whitespace:

it's defined as:
```ocaml
let isspace = function ' ' | '\n' | '\r' | '\t' -> true | _ -> false
```
